#MIT License
#
#Copyright (c) 2016 AUTHORS
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.#

diff -cr -N -x .git -x '*.patch' /old/AUTHORS /new/AUTHORS
*** /old/AUTHORS	1969-12-31 19:00:00.000000000 -0500
--- /new/AUTHORS	2016-12-21 19:15:05.011980080 -0500
***************
*** 0 ****
--- 1,2 ----
+ <<< user_name >>> <<< user_email >>> <<< user_signingkey >>>
+ 
diff -cr -N -x .git -x '*.patch' /old/CHANGELOG.md /new/CHANGELOG.md
*** /old/CHANGELOG.md	1969-12-31 19:00:00.000000000 -0500
--- /new/CHANGELOG.md	2016-12-21 19:15:04.975980081 -0500
***************
*** 0 ****
--- 1,16 ----
+ # Change Log
+ 
+ All notable changes to this project will be documented in this file.
+ This project adheres to [Semantic Versioning](http://semver.org/).
+ 
+ This Change Log format is suggested by
+ <https://github.com/olivierlacan/keep-a-changelog/blob/master/CHANGELOG.md>
+ 
+ <a name="0.0.1"></a>
+ ## 0.0.1 (<<< date >>>)
+ 
+ 
+ #### Features
+ 
+ *   Install and configure gitguild
+ *   Register patch
diff -cr -N -x .git -x '*.patch' /old/conf/gitolite.conf /new/conf/gitolite.conf
*** /old/conf/gitolite.conf	2016-12-22 14:13:38.792699998 -0500
--- /new/conf/gitolite.conf	2016-12-22 14:12:50.004700872 -0500
***************
*** 1,5 ****
  repo <<< user_name >>>
!     RW+     =   <<< user_name >>>
  
- repo testing
-     RW+     =   @all
--- 1,3 ----
  repo <<< user_name >>>
!     RW     =   <<< user_name >>>
  
diff -cr -N -x .git -x '*.patch' /old/conf/gitolite.conf~ /new/conf/gitolite.conf~
*** /old/conf/gitolite.conf~	1969-12-31 19:00:00.000000000 -0500
--- /new/conf/gitolite.conf~	2016-12-22 14:12:49.780700876 -0500
***************
*** 0 ****
--- 1,3 ----
+ repo <<< user_name >>>
+     RW     =   <<< user_name >>>
+ 
diff -cr -N -x .git -x '*.patch' /old/CONTRIBUTING.md /new/CONTRIBUTING.md
*** /old/CONTRIBUTING.md	1969-12-31 19:00:00.000000000 -0500
--- /new/CONTRIBUTING.md	2016-12-21 19:15:05.011980080 -0500
***************
*** 0 ****
--- 1,3 ----
+ # A GitGuild Project
+ 
+ This project uses [GitGuild](https://github.com/GitGuild/gitguild-cli) software to govern itself and all contributions. Please install the client to ensure you meet the strict configuration and usage requirements.
diff -cr -N -x .git -x '*.patch' /old/GUILD /new/GUILD
*** /old/GUILD	1969-12-31 19:00:00.000000000 -0500
--- /new/GUILD	2016-12-21 19:15:05.011980080 -0500
***************
*** 0 ****
--- 1,5 ----
+ NAME <<< guild_name >>>
+ HEIGHT <<< height >>>
+ LAST_TRANSACTION <<< last_transaction >>>
+ EPOCH_TIME <<< epoch >>>
+ 
diff -cr -N -x .git -x '*.patch' /old/ledger/account.ledger /new/ledger/account.ledger
*** /old/ledger/account.ledger	1969-12-31 19:00:00.000000000 -0500
--- /new/ledger/account.ledger	2016-12-21 19:15:04.975980081 -0500
***************
*** 0 ****
--- 1,44 ----
+ ; Account structure is quite complex in the GitGuild, creating a symmetry between this ledger and the git tree itself.
+ ; A number of complex and potentially long account names may be difficult to read, but they're quite organized.
+ ; The basic structure is as follows.
+ ;
+ ; username:user account:repository:branch:parent commit hash:metadata
+ ;
+ ; Most of these are probably self explanatory. User account is the usual suspects of accounting: Income, Expenses, Assets, Liabilities, Equity.
+ ; User account also has two additional options related to voting: Height, Depth.
+ ; Metadata might be bitcoin or dash addresses, and/or transaction hashes.
+ 
+ account <<< guild_name >>>
+     note The Guild's own account.
+ account <<< guild_name >>>:Expenses
+     note The Guild's Expenses, i.e. Asset redemptions, contribution pay origination
+ 
+ account <<< guild_name >>>:Depth
+     note the Guild's voting account. Votes in XP the amount of qualified Income earned by member(s).
+     assert commodity == "XP"
+ account <<< guild_name >>>:Height
+     note the Guild's vote counting area. Votes by members on commits related to this guild are recorded here.
+     assert commodity == "XP"
+ account <<< guild_name >>>:Equity
+ account <<< guild_name >>>:Income
+ account <<< guild_name >>>:Assets
+ account <<< guild_name >>>:Liabilities
+     note This guild can only have XGG liabilities, and this is the source of all XGG.
+ account <<< guild_name >>>:Expenses
+     note Members earning Income on approved labor can ask to cash in XGG against the Guild's Assets, creating an Expense
+ 
+ account <<< user_name >>>:Depth
+     note The votes that <<< user_name >>> has spent. Cannot spend more than account height on a commit/issue.
+ account <<< user_name >>>:Height
+     note The number of votes that <<< user_name >>> has received from the guild and also the XP voting power of <<< user_name >>>.
+ ; users do not get Equity as members of a shared guild.
+ account <<< user_name >>>:Income
+ account <<< user_name >>>:Assets
+ account <<< user_name >>>:Liabilities
+ account <<< user_name >>>:Expenses
+ 
+ account <<< user_name >>>:Equity:init:<<< parent >>>
+ 
+ account <<< user_name >>>:Equity:register:<<< parent >>>
+ 
+ 
diff -cr -N -x .git -x '*.patch' /old/ledger/chain.ledger /new/ledger/chain.ledger
*** /old/ledger/chain.ledger	1969-12-31 19:00:00.000000000 -0500
--- /new/ledger/chain.ledger	2016-12-21 19:15:04.975980081 -0500
***************
*** 0 ****
--- 1,6 ----
+ !include commodity.ledger
+ !include account.ledger
+ !include tag.ledger
+ !include equity.ledger
+ !include transaction.ledger
+ !include check.ledger
diff -cr -N -x .git -x '*.patch' /old/ledger/check.ledger /new/ledger/check.ledger
*** /old/ledger/check.ledger	1969-12-31 19:00:00.000000000 -0500
--- /new/ledger/check.ledger	2016-12-21 19:15:04.975980081 -0500
***************
*** 0 ****
--- 1,26 ----
+ ; TODO lots of assertions and checks go here to ensure proper behavior
+ ; these will be suplimented with file templates and transformations
+ 
+ define GUILD_DEPTH = (-account("<<< guild_name >>>:Depth:<<< guild_name >>>").total)
+ define GUILD_HEIGHT = (account("<<< guild_name >>>:Height:<<< guild_name >>>").total)
+ 
+ ; Each user must have net positive score
+ define user_vote_check(user_name) = (account(user_name + ":Height").total >= GUILD_DEPTH / 2)
+ 
+ ; Each branch of this guild must have net positive score
+ define branch_vote_check(user_name, vote_guild_name, vote_branch) = (account(user_name + ":Height:" + vote_guild_name + ":" + vote_branch).total >= GUILD_DEPTH / 2)
+ 
+ ; Each issue must have net positive score
+ define issue_vote_check(user_name, vote_guild_name, vote_branch, vote_parent) = (account(user_name + ":Height:" + vote_guild_name + ":" + vote_branch + ":" + vote_parent).total >= GUILD_DEPTH / 2)
+ 
+ ; user has not spent more XP than allowed on this issue
+ define user_issue_vote_check(user_name, vote_guild_name, vote_branch, vote_parent) = (account(user_name + ":Height").total >= abs(account(user_name + ":Depth:" + vote_guild_name + ":" + vote_branch + ":" + vote_parent).total))
+ 
+ ; Guild itself must have positive height
+ check user_vote_check("<<< guild_name >>>")
+ 
+ check branch_vote_check("<<< guild_name >>>", "<<< guild_name >>>", "<<< guild_name >>>")
+ check issue_vote_check("<<< guild_name >>>", "<<< guild_name >>>", "<<< guild_name >>>", "<<< vote_parent >>>")
+ check branch_vote_check("<<< guild_name >>>", "<<< guild_name >>>", "<<< guild_name >>>")
+ check user_issue_vote_check("<<< guild_name >>>", "<<< guild_name >>>", "<<< guild_name >>>", "<<< vote_parent >>>")
+ 
diff -cr -N -x .git -x '*.patch' /old/ledger/commodity.ledger /new/ledger/commodity.ledger
*** /old/ledger/commodity.ledger	1969-12-31 19:00:00.000000000 -0500
--- /new/ledger/commodity.ledger	2016-12-21 19:15:04.975980081 -0500
***************
*** 0 ****
--- 1,19 ----
+ commodity XP
+    note Experience Points, used for voting on members and issues
+    format 1,000.00 XP
+    nomarket
+ 
+ commodity XGG
+    note GitGuild token of value. Transferable and redeemable at bid rate if part of approved Income transaction.
+    format 1,000.0000 XGG
+    nomarket
+ 
+ commodity DASH
+    note See dash.org
+    format 1,000.00000000 DASH
+    nomarket
+ 
+ commodity BTC
+    note See bitcoin.org
+    format 1,000.00000000 BTC
+    nomarket
diff -cr -N -x .git -x '*.patch' /old/ledger/equity.ledger /new/ledger/equity.ledger
*** /old/ledger/equity.ledger	1969-12-31 19:00:00.000000000 -0500
--- /new/ledger/equity.ledger	2016-12-22 11:31:26.964874276 -0500
***************
*** 0 ****
--- 1 ----
+ # This file should have an 'equity' summary of all history up to the parent commit.
diff -cr -N -x .git -x '*.patch' /old/ledger/tag.ledger /new/ledger/tag.ledger
*** /old/ledger/tag.ledger	1969-12-31 19:00:00.000000000 -0500
--- /new/ledger/tag.ledger	2016-12-21 19:15:04.975980081 -0500
***************
*** 0 ****
--- 1,18 ----
+ tag init_personal
+ ; mv template/init/personal/ledger/equity.ledger ledger/equity.ledger
+ ; mv template/init/personal/ledger/commodity.ledger ledger/commodity.ledger
+ ; mv template/init/personal/ledger/account.ledger ledger/account.ledger
+ ; mv template/init/personal/ledger/tag.ledger ledger/tag.ledger
+ ; mv template/init/personal/ledger/transaction.ledger ledger/transaction.ledger
+ ; mv template/init/personal/ledger/chain.ledger ledger/chain.ledger
+ ; mv template/init/personal/ledger/check.ledger ledger/check.ledger
+ ; mv template/init/LICENSE LICENSE
+ ; mv template/init/AUTHORS AUTHORS
+ ; mv template/init/CONTRIBUTING.md CONTRIBUTING.md
+ ; mv template/init/CHANGELOG.md CHANGELOG.md
+ 
+ tag register_personal
+ ; patch template/register/ledger/account.ledger ledger/account.ledger
+ ; patch template/register/ledger/transaction.ledger ledger/transaction.ledger
+ ; patch template/init/personal/AUTHORS  AUTHORS
+ 
diff -cr -N -x .git -x '*.patch' /old/ledger/transaction.ledger /new/ledger/transaction.ledger
*** /old/ledger/transaction.ledger	1969-12-31 19:00:00.000000000 -0500
--- /new/ledger/transaction.ledger	2016-12-21 19:15:04.975980081 -0500
***************
*** 0 ****
--- 1,6 ----
+ 
+ <<< date >>> * <<< user_name >>> Personal Guild Initialization
+     ; Create XP by giving the founder Equity
+     <<< user_name >>>:Height:<<< user_name >>>:<<< user_name >>>:<<< parent >>>    1 XP
+     <<< user_name >>>:Depth:<<< user_name >>>:<<< user_name >>>:<<< parent >>>    -1 XP
+ 
diff -cr -N -x .git -x '*.patch' /old/VERSION /new/VERSION
*** /old/VERSION	1969-12-31 19:00:00.000000000 -0500
--- /new/VERSION	2016-12-22 11:36:33.260868791 -0500
***************
*** 0 ****
--- 1 ----
+ <<< version >>>
